query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/QueryRequestEligibilityValidator.java

```diff
@@ -0,0 +1,154 @@
+import org.hypertrace.core.query.service.api.QueryRequest;
+
+/** Set of rules to check if the given request can be served by prometheus */
+class QueryRequestEligibilityValidator {
+
+  private final PrometheusViewDefinition prometheusViewDefinition;
+
+  public QueryRequestEligibilityValidator(PrometheusViewDefinition prometheusViewDefinition) {
+    this.prometheusViewDefinition = prometheusViewDefinition;
+  }
+
+  QueryCost isEligible(QueryRequest queryRequest, ExecutionContext executionContext) {
+    // orderBy to be supported later
+    if (queryRequest.getOrderByCount() > 0) {
+      return QueryCost.UNSUPPORTED;
+    }
+
+    // only aggregation queries are supported
+    if (queryRequest.getAggregationCount() == 0
+        || queryRequest.getGroupByCount() == 0
+        || queryRequest.getDistinctSelections()) {
+      return QueryCost.UNSUPPORTED;
+    }
+
+    Set<String> referencedColumns = executionContext.getReferencedColumns();
+    Preconditions.checkArgument(!referencedColumns.isEmpty());
+    // all the columns in the request should have a mapping in the config
+    for (String referencedColumn : referencedColumns) {
+      if (!QueryRequestUtil.isTimeColumn(referencedColumn)
+          && prometheusViewDefinition.getPhysicalColumnName(referencedColumn) == null
+          && prometheusViewDefinition.getMetricConfig(referencedColumn) == null) {
+        return QueryCost.UNSUPPORTED;
+      }
+    }
+
+    if (!analyseAggregationColumns(queryRequest.getAggregationList())) {
+      return QueryCost.UNSUPPORTED;
+    }
+
+    if (!selectionAndGroupByOnSameColumn(
+        queryRequest.getSelectionList(), queryRequest.getGroupByList())) {
+      return QueryCost.UNSUPPORTED;
+    }
+
+    if (!analyseFilter(queryRequest.getFilter())) {
+      return QueryCost.UNSUPPORTED;
+    }
+
+    // value 1.0 so that prometheus is preferred over others
+    return new QueryCost(1.0);
+  }
+
+  private boolean selectionAndGroupByOnSameColumn(
+      List<Expression> selectionList, List<Expression> groupByList) {
+    Set<String> selections = Sets.newHashSet();
+    for (Expression expression : selectionList) {
+      if (!QueryRequestUtil.isSimpleColumnExpression(expression)) {
+        return false;
+      }
+      selections.add(getLogicalColumnName(expression));
+    }
+
+    for (Expression expression : groupByList) {
+      // skip datetime convert group by
+      if (QueryRequestUtil.isDateTimeFunction(expression)) {
+        continue;
+      }
+      if (!QueryRequestUtil.isSimpleColumnExpression(expression)) {
+        return false;
+      }
+      if (!selections.remove(getLogicalColumnName(expression))) {
+        return false;
+      }
+    }
+    return selections.isEmpty();
+  }
+
+  private String getLogicalColumnName(Expression expression) {
+    String logicalColumnName;
+    if (expression.getValueCase() == ValueCase.COLUMNIDENTIFIER) {
+      logicalColumnName = expression.getColumnIdentifier().getColumnName();
+    } else {
```

Comment:
Can we add exception for other value cases ?

Reply:
Done, wrapped the code inside the entry method of the class in try, catch and returning QueryCost.UNSUPPORTED on exception